pyomo/contrib/doe/fim_doe.py

```diff
@@ -642,29 +838,55 @@ def compute_FIM(self, design_values, mode='sequential_finite', FIM_store_name=No
             # produce the sensitivity for fixed variables
             zero_sens = np.zeros(len(self.param_name))
             # loop over measurement variables and their time points
-            for mname in self.measurement_variables:
-                for tim in measurement_accurate_time:
-                    # get the measurement name in the model
-                    measure_name = mname+'[0,'+str(tim)+']'
-                    measurement_names.append(measure_name)
-                    # get right line number in kaug results
-                    if self.discretize_model is not None:
-                        # for DAE model, some variables are fixed
-                        try:
+            for mname in self.flatten_measure_name:
+                if '_index_' in mname:
+                    measure_name = mname.split('_index_')[0]
+                    measure_index = mname.split('_index_')[1]
+                    for tim in measurement_accurate_time[mname]:
+                        # get the measurement name in the model
+                        measurement_name = measure_name+'[0,'+measure_index+','+str(tim)+']'
+                        measurement_names.append(measurement_name)
+                        # get right line number in kaug results
+                        if self.discretize_model is not None:
+                            # for DAE model, some variables are fixed
+                            try:
+                                kaug_no = col.index(measurement_name)
+                                measurement_index.append(kaug_no)
+                                # get right line of dsdp
+                                dsdp_extract.append(dsdp_array[kaug_no])
+                            except:
+                                if self.verbose:
+                                    print('The variable is fixed:', measurement_name)
+                                # for fixed variables, the sensitivity are a zero vector
+                                dsdp_extract.append(zero_sens)
+                        else:
+                            kaug_no = col.index(measure_name)
+                            measurement_index.append(kaug_no)
+                            # get right line of dsdp
+                            dsdp_extract.append(dsdp_array[kaug_no])
+                else:
+                    for tim in self.flatten_measure_timeset[mname]:
+                        # get the measurement name in the model
+                        measure_name = mname+'[0,'+str(tim)+']'
```

Comment:
Move this into the measurement class.

Reply:
Ask the measurement class to return the name for the Pyomo models. This will remove duplicate code here. It will also make it easier to test.